modules/code-editor/src/backend/utils.ts

```diff
@@ -0,0 +1,122 @@
+import _ from 'lodash'
+
+import { FileDefinition, FileTypes } from './definitions'
+import { FILENAME_REGEX } from './editor'
+import { EditorError } from './editorError'
+import { EditableFile, FilePermissions } from './typings'
+
+export const filterBuiltin = (files: EditableFile[]) => {
+  return files.filter(x => !x.content.includes('//CHECKSUM:'))
+}
+
+export const getFileLocation = (file: EditableFile): { folder: string; filename: string } => {
+  const fileDef: FileDefinition = FileTypes[file.type]
+  const { baseDir, upsertLocation, upsertFilename } = fileDef.ghost
+
+  const folder = (upsertLocation && upsertLocation(file)) || baseDir
+  const filename = (upsertFilename && upsertFilename(file)) || file.location
+
+  return { folder, filename }
+}
+
+export const assertValidJson = (content: string): boolean => {
+  try {
+    JSON.parse(content)
+    return true
+  } catch (err) {
+    throw new EditorError(`Invalid JSON file. ${err}`)
+  }
+}
+
+export const assertValidFilename = (filename: string) => {
+  if (!FILENAME_REGEX.test(filename)) {
+    throw new EditorError('Filename has invalid characters')
+  }
+}
+
+export const arePermissionsValid = (
+  def: FileDefinition,
+  editableFile: EditableFile,
+  permissions: FilePermissions,
+  actionType: 'read' | 'write'
+): boolean => {
+  const hasGlobalPerm = def.allowGlobal && permissions[`global.${def.permission}`][actionType]
+  const hasScopedPerm = def.allowScoped && permissions[`bot.${def.permission}`][actionType]
+
+  const isGlobalValid = def.allowGlobal && !editableFile.botId
+  const isScopedValid = def.allowScoped && !!editableFile.botId
+
+  return (hasGlobalPerm && isGlobalValid) || (hasScopedPerm && isScopedValid)
+}
+
+export const validateFilePayload = async (
+  editableFile: EditableFile,
+  permissions: FilePermissions,
+  currentBotId: string,
+  actionType: 'read' | 'write'
+) => {
+  const { name, botId, type, content, location } = editableFile
+
+  const def: FileDefinition = FileTypes[type]
+  if (!def) {
+    throw new EditorError(`Invalid file type "${type}", only ${Object.keys(FileTypes)} are allowed at the moment`)
+  }
+
+  if (botId && botId.length && botId !== currentBotId) {
+    throw new EditorError(`Can't perform modification on bot ${botId}. Please switch to the correct bot to change it.`)
+  }
+
+  if (!arePermissionsValid(def, editableFile, permissions, actionType)) {
+    throw new EditorError(`No permission`)
+  }
+
+  if (def.isJSON && content) {
+    assertValidJson(content)
+  }
+
+  if (def.validate) {
+    const result = await def.validate(editableFile)
+    if (result) {
+      throw new Error(result)
+    }
+  }
+
+  if (def.filenames && !def.filenames.includes(location)) {
+    throw new EditorError(`Invalid file name. Must match ${def.filenames}`)
+  }
+
+  assertValidFilename(name)
+}
+
+export const buildRestrictedProcessVars = () => {
```

Comment:
What's that for ? I mean is this used ?

Reply:
Yes, it shows autocomplete when you type `process.` in the code editor, also showing exposed variables and their current value